src/main/java/com/snowflake/kafka/connector/internal/PartitionBuffer.java

```diff
@@ -0,0 +1,92 @@
+package com.snowflake.kafka.connector.internal;
+
+import java.util.Collection;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+/**
+ * Abstract class which holds buffered data per partition including its start offset, end offset,
+ * size.
+ *
+ * <p>The getData() method returns the data specific to the implementation.
+ *
+ * <p>Buffer stores the converted records to Json format which can be temporary or long lived.
+ *
+ * <p>Temporary would mean the partition buffer is emptied when all records from {@link
+ * com.snowflake.kafka.connector.SnowflakeSinkTask#put(Collection)} are processed and respective API
+ * is called.
+ *
+ * <p>Long lived would mean the data in partition would stay across two put APIs since the buffer
+ * thresholds were not met.
+ *
+ * <p>Please check respective implementation class for more details.
+ *
+ * @param <T> Return type of {@link #getData()}
+ */
+public abstract class PartitionBuffer<T> {
+  private int numOfRecord;
```

Comment:
is numOfRecord = lastOffset-firstOffset? If yes, could we use it instead?

Reply:
If I recall correctly there can be gaps in offsets so we probably want to stick with numOfRecords